File: ios/RadioClient/RadioClient/Models/SongItem.swift
import Foundation

struct SongItem: Codable, Identifiable, Equatable {
    let id: Int
    let title: String
    let artist: String
    let album: String?
    let albumId: Int?
    let year: Int?
    let duration: Double?
    let fileFormat: String?
    let replaygainTrackGain: Double?

    enum CodingKeys: String, CodingKey {
        case id, title, artist, album
        case albumId = "album_id"
        case year, duration
        case fileFormat = "file_format"
        case replaygainTrackGain = "replaygain_track_gain"
    }

    var fileExtension: String {
        (fileFormat ?? "mp3").lowercased()
    }
}

struct PlayedSong: Identifiable {
    let id = UUID()
    let song: SongItem
    let playedAt: Date
    let skipped: Bool
}

struct SyncResponse: Codable {
    let download: [SongItem]
}

struct Channel: Codable, Identifiable, Equatable {
    let id: Int
    let name: String
    let yearMin: Int?
    let yearMax: Int?
    let genreGroup: String?
    let genre: String?
    let artist: String?

    enum CodingKeys: String, CodingKey {
        case id, name, genre, artist
        case yearMin = "year_min"
        case yearMax = "year_max"
        case genreGroup = "genre_group"
    }

    var subtitle: String {
        var parts: [String] = []
        if let min = yearMin, let max = yearMax {
            parts.append("\(min)–\(max)")
        } else if let min = yearMin {
            parts.append("\(min) and newer")
        } else if let max = yearMax {
            parts.append("up to \(max)")
        }
        if let g = genreGroup { parts.append(g) }
        if let g = genre { parts.append(g) }
        if let a = artist { parts.append(a) }
        return parts.isEmpty ? "All music" : parts.joined(separator: " · ")
    }
}

struct ChannelsResponse: Codable {
    let channels: [Channel]
}

File: android/RadioClient/app/src/main/java/com/example/radioclient/model/SongItem.kt
package com.example.radioclient.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant
import java.util.UUID

@Serializable
data class SongItem(
    val id: Int,
    val title: String,
    val artist: String,
    val album: String? = null,
    @SerialName("album_id") val albumId: Int? = null,
    val year: Int? = null,
    val duration: Double? = null,
    @SerialName("file_format") val fileFormat: String? = null,
    @SerialName("replaygain_track_gain") val replaygainTrackGain: Double? = null,
) {
    val fileExtension: String
        get() = fileFormat?.lowercase() ?: "mp3"
}

data class PlayedSong(
    val id: String = UUID.randomUUID().toString(),
    val song: SongItem,
    val playedAt: Instant,
    val skipped: Boolean,
)

@Serializable
data class SyncResponse(
    val download: List<SongItem>,
)

@Serializable
data class SyncRequestPlayed(
    val id: Int,
    @SerialName("played_at") val playedAt: String,
    val skipped: Boolean,
)

@Serializable
data class SyncRequestNowPlaying(
    val id: Int,
    @SerialName("started_at") val startedAt: String,
)

@Serializable
data class SyncRequest(
    val played: List<SyncRequestPlayed>,
    @SerialName("buffer_cache_mb") val bufferCacheMb: Int,
    @SerialName("now_playing") val nowPlaying: SyncRequestNowPlaying? = null,
)
